src/ptvsd/_vendored/pydevd/_pydevd_bundle/pydevd_process_net_command_json.py

```diff
@@ -460,15 +461,41 @@ def on_source_request(self, py_db, request):
         :param SourceRequest request:
         '''
         source_reference = request.arguments.sourceReference
-        content = ''
+        server_filename = None
+        content = None
+
         if source_reference != 0:
             server_filename = pydevd_file_utils.get_server_filename_from_source_reference(source_reference)
-            if os.path.exists(server_filename):
-                with open(server_filename, 'r') as stream:
-                    content = stream.read()
+            if server_filename:
```

Comment:
I think we really should have tests for these cases (error on `open` and using `linecache`).

Reply:
Testing `linecache` proved to be tricky - I tried a few approaches, but eventually had to resort to just adding entries directly to the cache. See the new commit.